contracts/Marketplace.sol

```diff
@@ -223,205 +193,92 @@ contract Marketplace is Collateral, Proofs, StateRetrieval {
-    if (_isCancelled(requestId)) {
+  function requestState(
+    RequestId requestId
+  ) public view requestIsKnown(requestId) returns (RequestState) {
+    RequestContext storage context = requestContexts[requestId];
+    if (
+      context.state == RequestState.New &&
+      block.timestamp > requests[requestId].expiry
+    ) {
       return RequestState.Cancelled;
-    } else if (_isFinished(requestId)) {
+    } else if (
+      context.state == RequestState.Started && block.timestamp > context.endsAt
+    ) {
       return RequestState.Finished;
     } else {
-      RequestContext storage context = _context(requestId);
       return context.state;
     }
   }
 
-  /// @notice returns true when the request is accepting proof submissions from hosts occupying slots.
-  /// @dev Request state must be new or started, and must not be cancelled, finished, or failed.
-  /// @param slotId id of the slot, that is mapped to a request, for which to obtain state info
-  function _slotAcceptsProofs(SlotId slotId) internal view returns (bool) {
-    RequestId requestId = _getRequestIdForSlot(slotId);
-    return _requestAcceptsProofs(requestId);
-  }
-
-  /// @notice returns true when the request is accepting proof submissions from hosts occupying slots.
-  /// @dev Request state must be new or started, and must not be cancelled, finished, or failed.
-  /// @param requestId id of the request for which to obtain state info
-  function _requestAcceptsProofs(
-    RequestId requestId
-  ) internal view returns (bool) {
-    RequestState s = state(requestId);
-    return s == RequestState.New || s == RequestState.Started;
-  }
-
-  function _notEqual(RequestId a, uint256 b) internal pure returns (bool) {
-    return RequestId.unwrap(a) != bytes32(b);
-  }
-
-  enum RequestState {
-    New, // [default] waiting to fill slots
-    Started, // all slots filled, accepting regular proofs
-    Cancelled, // not enough slots filled before expiry
-    Finished, // successfully completed
-    Failed // too many nodes have failed to provide proofs, data lost
+  function slotState(SlotId slotId) public view override returns (SlotState) {
+    Slot storage slot = slots[slotId];
+    if (RequestId.unwrap(slot.requestId) == 0) {
+      return SlotState.Free;
+    }
+    RequestState reqState = requestState(slot.requestId);
+    if (slot.state == SlotState.Paid) {
+      return SlotState.Paid;
+    }
+    if (reqState == RequestState.Cancelled) {
+      return SlotState.Finished;
+    }
+    if (reqState == RequestState.Finished) {
+      return SlotState.Finished;
+    }
+    if (reqState == RequestState.Failed) {
+      return SlotState.Failed;
+    }
+    return slot.state;
   }
 
   struct RequestContext {
-    uint256 slotsFilled;
     RequestState state;
+    uint256 slotsFilled;
     uint256 startedAt;
     uint256 endsAt;
   }
 
   struct Slot {
-    address host;
-    bool hostPaid;
+    SlotState state;
     RequestId requestId;
+    address host;
   }
```

Comment:
I'm a bit hesitant to do that, because I tend to think of `Requests.sol` as containing the definition of a request and its states (very close to a spec), and the `Slot` struct is just an incidental storage struct that happens to work for the Marketplace.